File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from collections import defaultdict

def solution(orders, course):
    answer = []
    ord = [list(o) for o in orders]
    
    for c in course:
        dic = defaultdict(int)
        
        for o in ord:
            def backtracking(idx, curr):
                if len(curr) == c:
                    dic["".join(sorted(curr))] += 1
                    return
            
                for i in range(idx, len(o)):
                    if o[i] not in curr:
                        curr.append(o[i])
                        backtracking(i+1, curr)
                        curr.pop()
            if len(o) >= c:
                backtracking(0, [])
        
        max_value = max(dic.values(), default=0)
        for k,v in dic.items():
            if v == max_value and v >= 2:
                answer.append(k)
    return sorted(answer)

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    Map<String,Integer> m = new HashMap<>();
    
    public void dfs(String order, int course, int st, Set<Integer> s) {
        if(s.size() == course) {
            List<Integer> list = new ArrayList<>(s);
            Collections.sort(list);
            StringBuilder sb = new StringBuilder();
            for(int index: list) sb.append(order.charAt(index));
            String result = sb.toString();
            m.put(result, m.getOrDefault(result,0) + 1);
            return;
        }
        
        for(int i=st;i<order.length();i++) {
            if(!s.contains(i)) {
                s.add(i);
                dfs(order, course, i+1, s);
                s.remove(i);
            }
        }
        
    }
    public String[] solution(String[] orders, int[] course) {
        int idx = 0;
        for(String order: orders) {
            char[] arr = order.toCharArray();
            Arrays.sort(arr);
            String sortedOrder = new String(arr);
            for (int c : course) {
                dfs(sortedOrder, c, 0, new HashSet<>());
            }
        }
        
        List<String> answer = new ArrayList<>();
        for(int c: course) {
            int max = 0;
            for(Map.Entry<String,Integer> en: m.entrySet()) {
                if(en.getKey().length() == c && en.getValue() >= 2) {
                    max = Math.max(max, en.getValue());
                }
            }
            
            for(Map.Entry<String,Integer> en: m.entrySet()) {
                if(en.getKey().length() == c && en.getValue() == max) {
                    answer.add(en.getKey());
                }
            }
        }
        Collections.sort(answer);
        return answer.toArray(new String[0]);
    }
}